File: KickStart/November_13_2021/Q2.cpp
#include <iostream>

int numStrokes(std::string painting) {
    int strokes = 0;
    bool prevRed = false;
    bool prevBlue = false;
    bool prevYellow = false;

    for (auto color : painting) {
        if (color == 'U') {
            if(prevBlue) {
                strokes++;
            }
            if(prevYellow) {
                strokes++;
            }
            if(prevRed) {
                strokes++;
            }
            prevRed = false;
            prevBlue = false;
            prevYellow = false;
        } else if (color == 'R') {
            if(prevBlue) {
                strokes++;
            }
            if (prevYellow) {
                strokes++;
            }
            prevRed = true;
            prevBlue = false;
            prevYellow = false;
        } else if (color == 'Y') {
            if(prevBlue) {
                strokes++;
            }
            if(prevRed) {
                strokes++;
            }
            prevRed = false;
            prevBlue = false;
            prevYellow = true;
        } else if (color == 'B') {
            if(prevRed) {
                strokes++;
            }
            if(prevYellow) {
                strokes++;
            }
            prevRed = false;
            prevBlue = true;
            prevYellow = false;
        } else if (color == 'O') {
            if(prevBlue) {
                strokes++;
            }
            prevRed = true;
            prevBlue = false;
            prevYellow = true;
        } else if (color == 'P') {
            if(prevYellow) {
                strokes++;
            }
            prevRed = true;
            prevBlue = true;
            prevYellow = false;
        } else if (color == 'G') {
            if (prevRed) {
                strokes++;
            }
            prevRed = false;
            prevBlue = true;
            prevYellow = true;
        } else if (color == 'A') {
            prevRed = true;
            prevBlue = true;
            prevYellow = true;
        }
    }
    if(prevBlue) {
        strokes++;
    }
    if(prevYellow) {
        strokes++;
    }
    if(prevRed) {
        strokes++;
    }
    return strokes;
}

int main() {
    int testCases;
    std::cin >> testCases;
    for (int i = 0; i < testCases; i++) {
        int paintingLength;
        std::cin >> paintingLength;
        std::string painting;
        std::cin >> painting;

        std::cout << "Case #" << i + 1 << ": " << numStrokes(painting) << "\n";
    }
}

File: KickStart/November_13_2021/Q2.rs
use std::io;

fn numStrokes(painting: &str) -> u32 {
    let mut strokes = 0;
    let mut prevRed = false;
    let mut prevBlue = false;
    let mut prevYellow = false;

    for c in painting.chars() {
        match c {
            'R' => {
                if prevBlue {
                    strokes += 1;
                }
                if prevYellow {
                    strokes += 1;
                }
                prevRed = true;
                prevBlue = false;
                prevYellow = false;
            }
            'Y' => {
                if prevRed {
                    strokes += 1;
                }
                if prevBlue {
                    strokes += 1;
                }
                prevRed = false;
                prevBlue = false;
                prevYellow = true;
            }
            'B' => {
                if prevRed {
                    strokes += 1;
                }
                if prevYellow {
                    strokes += 1;
                }
                prevRed = false;
                prevBlue = true;
                prevYellow = false;
            }
            'O' => {
                if prevBlue {
                    strokes += 1;
                }
                prevRed = true;
                prevBlue = false;
                prevYellow = true;
            }
            'P' => {
                if prevYellow {
                    strokes += 1;
                }
                prevRed = true;
                prevBlue = true;
                prevYellow = false;
            }
            'G' => {
                if prevRed {
                    strokes += 1;
                }
                prevRed = false;
                prevBlue = true;
                prevYellow = true;
            }
            'A' => {
                prevRed = true;
                prevBlue = true;
                prevYellow = true;
            }
            _ => {
                if prevRed {
                    strokes += 1;
                }
                if prevBlue {
                    strokes += 1;
                }
                if prevYellow {
                    strokes += 1;
                }
                prevRed = false;
                prevBlue = false;
                prevYellow = false;
            }
        }
    }
    if prevRed {
        strokes += 1;
    }
    if prevBlue {
        strokes += 1;
    }
    if prevYellow {
        strokes += 1;
    }
    return strokes;
}

fn main() -> io::Result<()> {
    let mut input = String::new();
    io::stdin().read_line(&mut input)?;
    let test_cases: u32 = input.trim().parse().unwrap();

    for test_case in 1..test_cases + 1 {
        let mut result = 0;
        let mut paintingLength: u32;
        input.clear();
        io::stdin().read_line(&mut input)?;
        paintingLength = input.trim().parse().unwrap();
        let mut painting = String::new();
        io::stdin().read_line(&mut painting)?;
        painting.pop();
        let result = numStrokes(&painting);
        println!("Case #{}: {}", test_case, result);
    }
    return Ok(());
}